### PR TITLE
fix (jkube-kit) : Add default restartPolicy in case of controller type Job (#1278)

### LINKED ISSUE #1278: Enricher with type Job does not generate mandatory spec.template.spec.restartPolicy
### Description
If you configure 

`<jkube.enricher.jkube-controller.type>job</jkube.enricher.jkube-controller.type>
`

, do not set any additional YAML fragment for the Job, and run `mvn k8s:resource`, the resulted Job resource will not have a `spec.template.spec.restartPolicy`. When trying to apply the resource (`mvn k8s:apply`) you will get a message with 

`This is required Policy: Required value: valid values: "OnFailure", "Never". `

### Info

* Eclipse JKube version : 1.6.0
* Maven version (`mvn -v`) :3.8.4
```

```
* Kubernetes / Red Hat OpenShift setup and version : Kubernetes v1.23/ Red Hat OpenShift 4.8

* If it's a bug, how to reproduce : Do not use any **Job fragment** under **src/main/jkube** and run mvn k8s:resource.

* Sample Reproducer Project : *[GitHub Clone URL]*https://github.com/ammbra/cat-adoption-batch. To reproduce it, please use the jkube branch https://github.com/ammbra/cat-adoption-batch/tree/jkube

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description
Fix #1278

Right now JKube doesn't generate resources with pod spec's
`restartPolicy` set. This works for most of the controller type
resources (e.g `Deployment`, `StatefulSet`, `DaemonSet`, `ReplicaSet`,
`ReplicationController` ); If not provided Kubernetes assumes it to be
`Always`.

But in case of `Job` it doesn't work and Kubernetes gives `422`. Add a
default restart policy in case of `Job`.

[kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have implemented unit tests to cover my changes
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java [hunk 1]
```diff
@@ -55,6 +55,7 @@
     public static final String CREATE_EXTERNAL_URLS = "jkube.createExternalUrls";
     public static final String JKUBE_DOMAIN = "jkube.domain";
     public static final String JKUBE_ENFORCED_REPLICAS = "jkube.replicas";
+    public static final String DEFAULT_JOB_RESTART_POLICY = "OnFailure";
```
Comment (manusa) [inline jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java:58]: Was this the default behavior when it was an optional field? do you have any reference?

Comment (rohanKanojia) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java:58]: As pointed out in PR description, the default value is `Always` (if not specified) for most controllers. However, for `Job` it's mandatory to provide restart policy

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

File: jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/DefaultControllerEnricher.java [hunk 2]
```diff
@@ -145,4 +146,10 @@ public void create(PlatformMode platformMode, KubernetesListBuilder builder) {
     return spec.getSpec().getContainers().stream().map(Container::getName).collect(Collectors.toList());
   }
 
+  private static String getDefaultContainerRestartPolicy(String controllerType) {
+    if (controllerType != null && controllerType.equalsIgnoreCase(ControllerType.JOB.name())) {
+      return DEFAULT_JOB_RESTART_POLICY;
+    }
+    return null;
+  }
```
Comment (manusa) [inline jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/DefaultControllerEnricher.java:154]: this should be delegated to the handler (it makes no sense having an if-else condition here, given that we have specific handlers to take care of specifics).

Comment (rohanKanojia) [reply jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/DefaultControllerEnricher.java:154]: JobHandler seems to be generating pod template spec using PodTemplateHandler:
https://github.com/eclipse/jkube/blob/b4c31fe43d6ee853bcf85128eb31727176a9d258/jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/JobHandler.java#L67-L70

We don't have any way of knowing whether PodTemplate is generated for Job or any other kind of controller in PodTemplateHandler. That's why I'm setting restart policy from DefaultControllerEnricher from where handlers are invoked.

File: jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java (hunk 1 above)
Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java]: But what was the behavior for Job controllers when this value wasn't required. That's the question

Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java]: Right now, we're assuming it is `OnFailure`.

Comment (rohanKanojia) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java]: It wasn't provided and Job was created without any `restartPolicy`. This used to get rejected by Kubernetes api server on apply phase

Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java]: I understand the issue. What I mean, is that I assumed that at some point this was working (however I'm doubting this more and more), so the kube API changed the requirements for Job Pod Templates at some point.

Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/BaseEnricher.java]: OK, so my worries are confirmed, this feature never worked...

https://github.com/kubernetes/kubernetes/issues/41194

**Review by manusa — APPROVED** [review]

**Pull request merged**